### ISSUE TITLE
`redundantStaticSelf` rule breaks build

### ISSUE THREAD

**Issue by calda** [nicklockwood/SwiftFormat#1446]
I noticed the new `redundantStaticSelf` rule on `develop`, which looks great (nice work @sjavora). I ran this new rule on our codebase and found that it removes `Self` in many cases where it isn't actually redundant, breaking the build. 

Here's one example:

```swift
public enum MyFeatureCacheStrategy {
  case networkOnly
  case cacheFirst

  public static let defaultCacheAge: TimeInterval = .minutes(5)

  public func requestStrategy<Outcome>() -> SingleRequestStrategy<Outcome> {
    switch self {
    case .networkOnly:
      return .networkOnly(writeResultToCache: true)
    case .cacheFirst:
      // `Self.` is unexpectedly removed, breaking the build
      return .cacheFirst(maxCacheAge: Self.defaultCacheAge)
    }
  }
}
```

I left a comment about this on the implementation here: https://github.com/nicklockwood/SwiftFormat/pull/1437/files#r1187656440

**Comment by sjavora** [comment]
Hey @calda, thanks for pointing this out - I was sure something like this was going to show up, which is why I wanted the rule not enabled by default 😅 

I'll try to update the implementation some time this week 👍

**Comment by calda** [comment]
I can help validate the new rule implementation against Airbnb's codebase. If it works on our (pretty huge) codebase without breaking the build then I think we can trust the implementation is safe enough to ship by default.

**Comment by electricalen** [comment]
@sjavora I have another case where there's a computed property inside a closure, for example:

`
var classProperty: String
method() { [weak self] var in

      guard let self else {
        return
      }

      var computedProperty: [String: Any] {
        [
          "item": classProperty
        ]
      }
`

**Comment by sjavora** [comment]
Ok, I could use some help - I tried using `mapRecursiveDeclarations`, but that didn't work. I _think_ the reason is that in that method, the "simple" declaration is handled as
```swift
case .declaration:
    return declaration
```
So the `mapped` value is thrown away in this case. I tried changing that code to 
```swift
case .declaration:
    return mapped
```
which yielded `Failed to disable redundantStaticSelf rule` when running tests. Not sure how that could happen - I doubt each rule is supposed to handle that by itself 🤔

**Comment by calda** [comment]
I factored out an `endOfDeclaration(at:) -> Int` helper from the implementation of `parseDeclarations()` that made it easy to implement a fix: https://github.com/nicklockwood/SwiftFormat/pull/1449

**Comment by sjavora** [comment]
Cheers for the fix 👍

**Comment by calda** [comment]
@sjavora: I confirmed that that with the fix in #1449 the `redundantStaticSelf` rule is applied to our codebase successfully and doesn't cause any build issues. Nice work! Thanks for this one.
